### PR TITLE
Add KeyringMode unit property to fix cryptsetup key caching

### PR DESCRIPTION
**Pull request by poettering**
Mostly an implementation of what was proposed here:

https://github.com/systemd/systemd/pull/6286#issuecomment-314466259

I am not sure if this fixes all the key caching issues reported. @eworm-de can I bribe you into testing this PR and checking if this fixes the caching issues for you properly?

### CODE REVIEW

File: man/systemd.exec.xml [hunk 1]
```diff
@@ -1479,7 +1479,7 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
               </row>
               <row>
                 <entry>@keyring</entry>
-                <entry>Kernel keyring access (<citerefentry project='man-pages'><refentrytitle>keyctl</refentrytitle><manvolnum>2</manvolnum></citerefentry> and related calls)</entry>
+                <entry>Kernel key ring access (<citerefentry project='man-pages'><refentrytitle>keyctl</refentrytitle><manvolnum>2</manvolnum></citerefentry> and related calls)</entry>
```
Comment (keszybz) [inline man/systemd.exec.xml]: "keyring" is a valid word nowadays.

File: src/core/execute.c [hunk 2]
```diff
@@ -2207,6 +2217,55 @@ static int setup_keyring(Unit *u, const ExecParameters *p, uid_t uid, gid_t gid)
                 if (keyctl(KEYCTL_CHOWN, keyring, uid, gid, 0) < 0)
                         return log_error_errno(errno, "Failed to change ownership of session keyring: %m");
 
+        /* When requested link the user keyring into the session keyring. */
+        if (context->key_ring_mode == EXEC_KEY_RING_SHARED) {
+                uid_t saved_uid;
+                gid_t saved_gid;
+
+                /* Acquiring a reference to the user keyring is nasty. We briefly change identity in oder to get things
```
Comment (keszybz) [inline src/core/execute.c]: typo

File: man/systemd.exec.xml (hunk 1 above)
Comment (poettering) [reply man/systemd.exec.xml]: hmm, i checked merriem webster specifically, and it insisted on "key ring"... But i figure if it's "keyring" according to wikipedia in respect to cryptography, i figure that's fine too

Comment (keszybz) [reply man/systemd.exec.xml]: Yeah, I think it'd be confusing. Internally in the code we have `keyring` in various places, so adding `key_ring` in others is inconsistent.

**Comment by poettering** [comment]
force pushed a new version that makes the KeyRing → Keyring change, as @keszybz suggested (and the typo fix too).

Please have another look!

**Comment by keszybz** [comment]
Whoa, isn't this missing some .gperf changes?

**Comment by poettering** [comment]
hmm, indeed. not sure where that commit got lost

**Comment by poettering** [comment]
somehow the unit file change got lost in my rebase orgy. Readded that stuff from scracth now. Please have a look.

**Comment by keszybz** [comment]
`× xenial-i386` fails in `networkd-test.py` again, most likely unrelated.

**Comment by keszybz** [comment]
I booted my notebook with this, everything seems to work nicely.

**Comment by eworm-de** [comment]
Looks good, thanks a lot!
My code works with just little modifications.

File: src/core/execute.c [hunk 3]
```diff
@@ -2213,6 +2223,55 @@ static int setup_keyring(Unit *u, const ExecParameters *p, uid_t uid, gid_t gid)
...
+                        }
+                }
+
+                if (keyctl(KEYCTL_LINK,
+                           KEY_SPEC_USER_KEYRING,
+                           KEY_SPEC_SESSION_KEYRING, 0, 0) < 0) {
+
+                        r = -errno;
+
+                        (void) setreuid(saved_uid, -1);
```
Comment (setharnold) [inline src/core/execute.c:2256]: The setreiud(2) manpage includes a warning against calling setreuid(2) without checking the error return:

```
       Note: there are cases where setreuid() can fail even when the
       caller is UID 0; it is a grave security error to omit
       checking for a failure return from setreuid().
```

Comment (poettering) [reply src/core/execute.c:2256]: Well, this is a fatal error path already. Generally, when an error path itself results in another error you're fucked. In this case we do the best we can, and make sure to return the first error that happens and the caller will eventually exit due to this.

**Pull request merged**